added migration guide for Bolt@2.x (#442)
slackapi/bolt: docs/_tutorials/migration_v2.md

---
title: Migrating to V2
order: 1
slug: migration-v2
lang: en
layout: tutorial
permalink: /tutorial/migration-v2
---
# Migrating to v2.x

<div class="section-content">

> This guide will walk you through the process of updating your app from using `bolt@v1.x` to `bolt@2.x`. 

Reviewer: ```suggestion
This guide will walk you through the process of updating your app from using `@slack/bolt@v1.x` to `@slack/bolt@2.x`. 
```

Reviewer: Maybe a sentence like "There are a few changes you'll need to make but for most apps, these changes can be applied in 5 - 15 minutes."

</div> 

---

## Upgrading your listeners to `async`

Listeners in your app should updated to `async` functions and `say()`,  `respond()`, and `ack()` should be prefaced with `await`.

Before:

```javascript
app.action('some-action-id', ({action, ack, say}) => { 
  ack();
  say('hello world');
})
```

After:

```javascript
app.action('some-action-id', async ({action, ack, say}) => { 
  await ack();
  await say('hello world');
})
```

## Error Handling

> The recent changes in Bolt for Javascript V2 have improved our ability to catch errors and filter them to the global error handler.  It is still recommended to manage errors in the listeners themselves instead of letting them propagate to the global handler when possible.

Reviewer 2: nit: s/Javascript/JavaScript/**g**

### Handling Errors in Listeners with `try`/`catch`

```javascript
app.action('some-action-id', async ({action, ack, say}) => { 
  try {
    await ack();
    await say('hello world');
  } catch (error) {

>     console.log(error)

Reviewer 2: Let's use `logger` instead.
```js
async ({ logger, ... }) => {
  try {
  } catch (error) {
    logger.error(error);
  }
}
```

    // handle error
  }
})
```

### Handling Errors with the Global Error Handler

```javascript
app.error((error) => {
  // Check the details of the error to handle cases where you should retry sending a message or stop the app

>   console.error(error);

Reviewer 2: By contrast, there are no other ways apart from using `console` directly here.

});
```

Other error related changes include:

> - When your listener doesn’t call `ack` within the 3 section time limit, we log the failure instead of throwing an error.

Reviewer 2: 3 second time limit?

> - If you have multiple errors at once when running middleware, Bolt for Javascript will return a wrapper error with a `code` parameter of `slack_bolt_multiple_listener_error` and an `original` parameter that contains an array of all of the errors. 

Reviewer: We don't want people using the string value of error codes directly. It's much more future-proof if they use the enums. Also fix on `original` to `originals` (plural).

```suggestion
- If multiple errors occur when processing multiple listeners for a single event, Bolt for Javascript will return a wrapper error with a `code` property of `ErrorCode.MultipleListenerError` and an `originals` property that contains an array of the individual errors. 
```

## Message Shortcuts

[Message shortcuts](https://api.slack.com/interactivity/shortcuts/using#message_shortcuts) (previously referred to as message actions) now use the `shortcut()` method instead of the `action()` method.

Before:

```javascript

> app.action('message-action-callback', ({action, ack, context}) => {

Reviewer: ```suggestion
app.action({ callback_id: 'message-action-callback' }, ({action, ack, context}) => {
```
